overlord/devicestate/devicemgr.go

```diff
@@ -2214,3 +2214,140 @@ func (m *DeviceManager) RemoveRecoveryKeys() error {
+	StorageSafety asserts.StorageSafety
+	// Available is set to true if encryption is available on this device
+	// with the used gadget.
+	Available bool
+
+	// Type us set to the EncryptionType that can be used if
+	// Available is true.
+	Type secboot.EncryptionType
+
+	// UnvailableErr is set if the encryption support availability of
+	// the this device and used gadget do not match the
+	// storage safety requirements.
+	UnavailableErr error
+	// UnavailbleWarning describes why encryption support is not
+	// available in case it is optional.
+	UnavailableWarning string
+}
+
+var secbootCheckTPMKeySealingSupported = secboot.CheckTPMKeySealingSupported
+
+// checkEncryption verifies whether encryption should be used based on the
+// model grade and the availability of a TPM device or a fde-setup hook
+// in the kernel.
+func (m *DeviceManager) checkEncryption(st *state.State, deviceCtx snapstate.DeviceContext) (secboot.EncryptionType, error) {
+	model := deviceCtx.Model()
+
+	kernelInfo, err := snapstate.KernelInfo(st, deviceCtx)
+	if err != nil {
+		return "", fmt.Errorf("cannot check encryption support: %v", err)
+	}
+	var gadgetInfo *gadget.Info
+	if gadgetSnapInfo, err := snapstate.GadgetInfo(st, deviceCtx); err == nil {
+		gadgetInfo, err = gadget.ReadInfo(gadgetSnapInfo.MountDir(), nil)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	res, err := m.encryptionSupportInfo(model, kernelInfo, gadgetInfo)
+	if err != nil {
+		return "", err
+	}
+	if res.UnavailableWarning != "" {
+		logger.Noticef("%s", res.UnavailableWarning)
+	}
+	// encryption disabled or preferred unencrypted: follow the model preferences here even if encryption would be available
+	if res.Disabled || model.StorageSafety() == asserts.StorageSafetyPreferUnencrypted {
+		res.Type = secboot.EncryptionTypeNone
+	}
+
+	return res.Type, res.UnavailableErr
+}
+
+func (m *DeviceManager) encryptionSupportInfo(model *asserts.Model, kernelInfo *snap.Info, gadgetInfo *gadget.Info) (EncryptionSupportInfo, error) {
+	secured := model.Grade() == asserts.ModelSecured
+	dangerous := model.Grade() == asserts.ModelDangerous
+	encrypted := model.StorageSafety() == asserts.StorageSafetyEncrypted
+
+	res := EncryptionSupportInfo{
+		StorageSafety: model.StorageSafety(),
+	}
+
+	// check if we should disable encryption non-secured devices
+	// TODO:UC20: this is not the final mechanism to bypass encryption
+	if dangerous && osutil.FileExists(filepath.Join(boot.InitramfsUbuntuSeedDir, ".force-unencrypted")) {
+		res.Disabled = true
+	}
+
+	// check encryption: this can either be provided by the fde-setup
+	// hook mechanism or by the built-in secboot based encryption
+	hasFDESetupHook := hasFDESetupHookInKernel(kernelInfo)
+	// Add a default error here to ensure that even if the code is
+	// refactored we never return a nil checkEncryptionErr result
+	// even if no encryption check was done.
+	checkEncryptionErr := fmt.Errorf("internal error: no encryption check performed")
```

Comment:
Forgive me my naivity here, but I still don't get why this is done - if `checkEncryptionErr` is assigned in both branches, and both branches could return a nil result, then why is this neccessary?

Reply:
I may have overdone it here - today this is fine and the if/else will ensure that checkEncryptionErr is always set. However if would ever refactor this code and it becomes a "if(foo) else if (bar)" we may lose this property and silently suddently no checker is run and we don't notice. But again, maybe I'm overthinking it.

Reply:
Would it make more sense to add a unit test that can do this instead?

Reply:
I updated the code now and hopefully it's more clear and less confusing.